### ISSUE TITLE
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op

### ISSUE THREAD

**Issue by RangeLiu** [pytorch/pytorch#73843]
### 🐛 Describe the bug

When converting my .pth model to onnx, I got the following warning.
`Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.`

My converting codes:
`
 # set the model to inference mode
    model.eval()
    # Let's create a dummy input tensor
    dummy_input = torch.autograd.Variable(
        torch.randn(1, 3, input_shape[0], input_shape[1]))
    # Export the model
    torch.onnx.export(model,  
                      dummy_input,
                      output_file, 
                      verbose=True,
                      keep_initializers_as_inputs=True,
                      opset_version=11,  
                      input_names=['data'],
                      output_names=['output'],
                      )
`

The warned code line:
`x = F.pad(x, (0, 1, 0, 1), "constant", 0)`

Does opset11 not support constant padding? Thanks.

### Versions

PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 IoT 企业版
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.17.2
Libc version: N/A

Python version: 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 23:03:10) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.18362-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

Versions of relevant libraries:
[pip3] numpy==1.19.3
[pip3] torch==1.7.1
[pip3] torchaudio==0.7.2
[pip3] torchvision==0.8.2
[conda] Could not collect

**Comment by ZiyuBao** [comment]
same question. if i use opset 11, pad will be unfolded and the slice op. in it will not be quantizable. any solution please? if i use opset 9, no problem when converting to onnx and use onnx to infer. however, quantizing it using snpe can result in segmentation fault

**Comment by garymm** [comment]
@RangeLiu it's just a warning.
Can you please provide full repro code, and explain the expected vs observed behavior?
If you can provide code that has an assertion about the expected behavior that fails that'd be very useful.

@ZiyuBao sounds like your issues are probably best directed to whatever software is doing the quantization.

**Comment by dzenanz** [comment]
I am running into the same warning, except I pass `do_constant_folding=False` to `torch.onnx.export()`. I have described the issue in more detail on [discuss.pytorch.org](https://discuss.pytorch.org/t/how-to-suppress-constant-folding-warning/163756).

**Comment by shubhambhokare1** [comment]
@dzenanz Can you provide a sample repro?

**Comment by dzenanz** [comment]
```py
import os
import numpy
import torch
import segmentation_models_pytorch as smp

def save_model(model, output_path, dummy_input=None):
    """ Save the model at the end of training session both as pytorch and as onnx file.
        Give the dummy_input to save the onnx model also. If not present than only the pytorch
        model is saved.
    """

    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    # switch to eval mode
    model.eval()
    model_path = f"{output_path}.pth"
    torch.save(model.state_dict(), model_path)

    if dummy_input is not None:
        # also saving as onnx file
        model_path = f"{output_path}.onnx"
        # This is necessary to make sure that the model can be saved as onnx
        try:
            model.module.encoder.set_swish(memory_efficient=False)
        except AttributeError:
            model.encoder.set_swish(memory_efficient=False)
        dummy_input = dummy_input.cuda()
        torch.onnx.export(
            model,
            dummy_input,
            model_path,
            export_params=True,
            opset_version=11,
            do_constant_folding=False,
            input_names=["input"],
            output_names=["output"],
            dynamic_axes={"input": {0: "batch_size"}, "output": {0: "batch_size"}},
        )


if __name__ == "__main__":
    dummy_input = numpy.random.rand(4, 3, 368, 640).astype(numpy.float32)
    aux_params = {"dropout": 0.0, "classes": 3}
    model = smp.DeepLabV3Plus(
        "efficientnet-b0",
        in_channels=3,
        classes=2,
        decoder_channels=256,
        aux_params=aux_params,
    )
    model = model.cuda()
    save_model(model, "./test", torch.tensor(dummy_input))
```
`save_model` function is straight from our code. The invocation in main is minimal repro. Invocation result:
```log
C:\Users\Dzenan\miniconda3\envs\deep_learning\python.exe "C:/Program Files/JetBrains/PyCharm Community Edition 2021.1.1/plugins/python-ce/helpers/pydev/pydevd.py" --multiprocess --qt-support=auto --client 127.0.0.1 --port 52755 --file path/to/warningRepro.py 
Connected to pydev debugger (build 222.4345.23)
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.
Warning: Constant folding - Only steps=1 can be constant folded for opset >= 10 onnx::Slice op. Constant folding not applied.

Process finished with exit code 0
```
Here is our `environment.yml`:
```yaml
name: deep_learning
channels:
  - pytorch
  - conda-forge
  - defaults
dependencies:
  - python=3.8
  - pytorch=1.11.0
  - cudatoolkit=11.3.1
  - ffmpeg>=4.3.2
  - x264=1!161.3030
  - opencv=4.5.5
  - openh264>=2.1.1
  - matplotlib=3.5.1
  - numpy=1.22.3
  - pip=22.0.4
  - scikit-image=0.19.2
  - setuptools=60.10.0
  - tensorboard=2.8.0
  - tifffile=2022.3.16
  - jupyterlab=3.3.2
  - jupyter_contrib_nbextensions=0.5.1
  - jupytext=1.13.7
  - segmentation-models-pytorch=0.2.1
  - onnx=1.11.0
  - joblib=1.1.0
  - seaborn=0.11.2
  - openpyxl=3.0.9
  - pytest=7.1.1
  - pylint=2.12.2
  - scikit-learn=1.1.2
  - ipywidgets=7.6.5
  - cudnn=8.2.1
  - fpdf=1.7.2
  - natsort=7.1.1
  - pip:
    - monai==0.7.0
    - statannotations==0.4.4
    - tensorflow==2.8.0
    - tf2onnx==1.9.3
    - onnxruntime-gpu==1.10.0
    - h5py==3.7.0
    - pytorch-ignite==0.4.9
    - boto3==1.24.50
    - dataframe_image==0.1.1
    - itk==5.2.1.post1
    - tqdm==4.64.0
```
The warnings are also triggered on Ubuntu (don't remember whether 22.04 or 20.04).

**Comment by dzenanz** [comment]
Oh, and adding
```py
import warnings

warnings.filterwarnings(
    "once",
    message="Constant folding not applied",
)
```
to the `warningRepro.py` file does nothing.

**Comment by dzenanz** [comment]
Did you try my repro code?

**Comment by dzenanz** [comment]
Any progress?

**Comment by isaacraf95** [comment]
any updates?

**Comment by jcowles** [comment]
I'm hitting this problem as well

**Comment by JohnHerry** [comment]
same problem

**Comment by sesevasa64** [comment]
Same issue

**Comment by zhs108** [comment]
+1

**Comment by SchweitzerGAO** [comment]
Oh I am not alone...

**Comment by TheMalgot** [comment]
... neither am I.

The model I'm exporting can safely be used to infer, but takes twice the time than a similar Tensorflow implementation. I'm wondering if the missing folding is responsible for that.

**Comment by jeeprhymeallthetime** [comment]
+1

**Comment by ankushj8** [comment]
+1

**Comment by ar-roy** [comment]
+1

**Comment by Valerianding** [comment]
+1

**Comment by lix19937** [comment]
+1

**Comment by lix19937** [comment]
also https://discuss.pytorch.org/t/how-to-suppress-constant-folding-warning/163756

**Comment by OverLordGoldDragon** [comment]
@dzenanz `warnings` operates on regex: `warnings.filterwarnings("once", message=".*Constant folding not applied*")`.

**Comment by OverLordGoldDragon** [comment]
Still an issue in 2.3.0. MRE:

```python
import torch

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(2, 4, 3)

    def forward(self, x):
        x = torch.nn.functional.pad(x, (1, 1, 1, 1))
        return self.conv(x)

torch.onnx.export(Model(), torch.randn(1, 2, 9, 9), 'm.onnx')
```

**Comment by lix19937** [comment]
> @dzenanz `warnings` operates on regex: `warnings.filterwarnings("once", message=".*Constant folding not applied.*")`.

This is not the root solution.

**Comment by Kaze816** [comment]
这个问题还是没有解决呀？大佬们都很忙哦......
